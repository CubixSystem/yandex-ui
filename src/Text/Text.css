.Text {
    font-weight: var(--typography-font-weight-regular);

    color: var(--color-typo-primary);
}

.Text_align_start {
    text-align: left;
}

.Text_align_center {
    text-align: center;
}

.Text_align_end {
    text-align: right;
}

.Text_align_justify {
    text-align: justify;
}

.Text_overflow_fade,
.Text_overflow_fade-horizontal {
    position: relative;

    display: block;
    overflow: hidden;
}

.Text_overflow_fade::after,
.Text_overflow_fade-horizontal::after {
    position: absolute;
    right: 0;
    bottom: 0;

    content: '\00a0';
}

.Text_overflow_fade::after {
    width: 2em;

    background-image: linear-gradient(270deg, var(--color-bg-default) 35%, transparent);
}

.Text_overflow_fade-horizontal::after {
    left: 0;

    background-image: linear-gradient(0deg, var(--color-bg-default) 35%, transparent);
}

.Text_overflow_ellipsis {
    display: -webkit-box;
    -webkit-box-orient: vertical;
    overflow: hidden;

    white-space: normal;
    text-overflow: ellipsis;
}

.Text_color_brand {
    color: var(--color-typo-brand);
}

.Text_color_inverse {
    color: var(--color-typo-inverse);
}

.Text_color_promo {
    color: var(--color-typo-promo);
}

.Text_color_secondary {
    color: var(--color-typo-secondary);
}

.Text_color_ghost {
    color: var(--color-typo-ghost);
}

.Text_color_success {
    color: var(--color-typo-success);
}

.Text_color_alert {
    color: var(--color-typo-alert);
}

.Text_color_disable {
    color: var(--color-typo-disable);
}

.Text_color_warning {
    color: var(--color-typo-warning);
}

.Text_color_control-primary {
    color: var(--color-control-typo-primary);
}

.Text_color_control-secondary {
    color: var(--color-control-typo-secondary);
}

.Text_color_control-passive {
    color: var(--color-control-typo-passive);
}

.Text_color_control-ghost {
    color: var(--color-control-typo-ghost);
}

.Text_color_control-faint {
    color: var(--color-control-typo-faint);
}

.Text_color_control-disable {
    color: var(--color-control-typo-disable);
}

.Text_color_control-link {
    color: var(--color-control-typo-link);
}

.Text_color_control-error {
    color: var(--color-control-typo-error);
}

.Text_color_link {
    color: var(--color-link);
}

.Text_color_link-external {
    color: var(--color-link-external);
}

.Text_color_link-minor {
    color: var(--color-link-minor);
}

.Text_color_link-hover {
    color: var(--color-link-hover);
}
