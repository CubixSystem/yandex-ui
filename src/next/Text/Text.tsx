import React, { CSSProperties, Ref, useState, useEffect, useRef, WeakValidationMap, ValidationMap } from 'react';
import { Composition } from '@bem-react/core';
import { cn } from '@bem-react/classname';
import { canUseDOM } from '../lib/canUseDOM';
import { mergeAllRefs } from '../lib/mergeRefs';
import type { TextOverflowValue } from './_overflow/Text_overflow';

type ComponentType = React.ComponentType<{
    ref?: Ref<HTMLElement>;
    style?: CSSProperties;
    className?: string;
}>;

export type TextElementType = keyof React.ReactHTML | ComponentType;

/**
 * Упрощенный ElementType для обхода
 * "Expression produces a union type that is too complex to represent. ts(2590)"
 */
type ShortElementType = 'a' | 'td' | 'span' | ComponentType;

type CommonProps = {
    /**
     * Задает отображение переполненного текста
     */
    overflow?: TextOverflowValue;

    /**
     * Максимальное количество строк текста (работает в связке с overflow)
     * @default 1
     */
    maxLines?: number;

    /**
     * Задает типографику текста в компоненте.
     */
    typography?: unknown;
}

type Props<T extends TextElementType> = CommonProps & {
    ref?: T extends keyof JSX.IntrinsicElements
        ? (JSX.IntrinsicElements[T] extends React.ClassAttributes<infer R> ? Ref<R> : never)
        : (T extends React.ComponentType<{ ref?: infer R }> ? R : never);

    /**
     * Тип элемента для отображения как (строка или компонент).
     * @default "span"
     */
    as?: T;
};

export type TextProps<T extends TextElementType> =
    Props<T> & Omit<React.ComponentProps<T>, keyof Props<T>>;

export const cnText = cn('Text');

const getLineHeight = (textElement: HTMLElement) => {
    return parseFloat(getComputedStyle(textElement).lineHeight);
};

const isFade = (overflow: TextOverflowValue) => {
    return overflow === 'fade' || overflow === 'fade-horizontal';
};

const isEllipsis = (overflow: TextOverflowValue) => {
    return overflow === 'ellipsis';
};

export type Text<U = {}> = {
    <T extends TextElementType = 'span'>(props: TextProps<T> & U): React.ReactElement | null;
    propTypes?: WeakValidationMap<CommonProps & U>;
    contextTypes?: ValidationMap<any>;
    defaultProps?: Partial<CommonProps & U>;
    displayName?: string;
}

export type PropsOfText<C extends Text<any>, T extends TextElementType = 'span'> =
    C extends Text<infer U> ? TextProps<T> & U : never;

/**
 * Базовый примитив представления текстовых данных.
 * @param { TextProps } props
 */
export const Text = React.memo(<T extends ShortElementType = 'span'>(props: TextProps<T>) => {
    const {
        as: As = 'span',
        ref,
        className,
        overflow,
        maxLines = 1,
        style = {},
        typography,
        ...otherProps
    } = props;
    const [, forceUpdate] = useState({});
    const textElement = useRef<HTMLElement>(null);

    useEffect(() => {
        if (textElement && overflow && isFade(overflow) && canUseDOM()) {
            forceUpdate({});
        }
    }, [typography, overflow]);

    const newStyle = { ...style };
    if (overflow) {
        if (isEllipsis(overflow)) {
            newStyle.WebkitLineClamp = maxLines;
        } else if (isFade(overflow)) {
            newStyle.whiteSpace = maxLines === 1 ? 'nowrap' : undefined;
            if (maxLines > 1) {
                const lineHeight = textElement.current && getLineHeight(textElement.current);
                if (typeof lineHeight === 'number' && !isNaN(lineHeight)) {
                    newStyle.maxHeight = `${maxLines * lineHeight}px`;
                }
            }
        }
    }

    return (
        <As
            ref={mergeAllRefs(ref, textElement)}
            style={newStyle}
            className={cnText({}, [className])}
            {...otherProps}
        />
    );
}) as Text;
Text.displayName = cnText();

export function enhanceText<H extends Composition<any>, P = {}>(
    hoc: H, Comp?: Text<P>,
): Text<P & (H extends Composition<infer T> ? T : never)> {
    return hoc(Comp || Text) as any;
}
