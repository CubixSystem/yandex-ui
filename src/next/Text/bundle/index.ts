export * from '../Text.bundle/common';
