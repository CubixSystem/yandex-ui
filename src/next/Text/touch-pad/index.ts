export * from '../Text@touch-pad';
export * from '../_typography/Text_typography';
export * from '../_typography/Text_typography_body-long-l';
export * from '../_typography/Text_typography_body-long-m';
export * from '../_typography/Text_typography_body-long-s';
export * from '../_typography/Text_typography_body-long-xl';
export * from '../_typography/Text_typography_body-short-l';
export * from '../_typography/Text_typography_body-short-m';
export * from '../_typography/Text_typography_body-short-s';
export * from '../_typography/Text_typography_body-short-xl';
export * from '../_typography/Text_typography_caption-l';
export * from '../_typography/Text_typography_caption-m';
export * from '../_typography/Text_typography_caption-xl';
export * from '../_typography/Text_typography_control-l';
export * from '../_typography/Text_typography_control-m';
export * from '../_typography/Text_typography_control-s';
export * from '../_typography/Text_typography_control-xl';
export * from '../_typography/Text_typography_control-xs';
export * from '../_typography/Text_typography_control-xxl';
export * from '../_typography/Text_typography_control-xxs';
export * from '../_typography/Text_typography_display-l';
export * from '../_typography/Text_typography_display-m';
export * from '../_typography/Text_typography_display-s';
export * from '../_typography/Text_typography_display-xl';
export * from '../_typography/Text_typography_headline-l';
export * from '../_typography/Text_typography_headline-m';
export * from '../_typography/Text_typography_headline-s';
export * from '../_typography/Text_typography_headline-xl';
export * from '../_typography/Text_typography_headline-xs';
export * from '../_typography/Text_typography_overline-l';
export * from '../_typography/Text_typography_overline-m';
export * from '../_typography/Text_typography_overline-s';
export * from '../_typography/Text_typography_subheader-l';
export * from '../_typography/Text_typography_subheader-m';
export * from '../_typography/Text_typography_subheader-s';
export * from '../_typography/Text_typography_subheader-xl';
export * from '../_weight/Text_weight';
export * from '../_weight/Text_weight_bold';
export * from '../_weight/Text_weight_light';
export * from '../_weight/Text_weight_medium';
export * from '../_weight/Text_weight_regular';
export * from '../_color/Text_color';
export * from '../_overflow/Text_overflow';
export * from '../_align/Text_align';
