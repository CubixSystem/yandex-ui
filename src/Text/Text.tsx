import React, { ElementType, ReactNode, CSSProperties, PureComponent } from 'react';
import { cn } from '@bem-react/classname';
import { canUseDOM } from '../lib/canUseDOM';

import './Text.css';

export type TextAlignValue = 'start' | 'end' | 'center' | 'justify';

export type TextOverflowValue = 'fade' | 'ellipsis' | 'fade-horizontal';

export type TextColorValue =
    | 'brand'
    | 'inverse'
    | 'primary'
    | 'promo'
    | 'secondary'
    | 'ghost'
    | 'control-primary'
    | 'alert'
    | 'warning'
    | 'disable'
    | 'success'
    | 'control-secondary'
    | 'control-passive'
    | 'control-ghost'
    | 'control-faint'
    | 'control-disable'
    | 'control-link'
    | 'control-error'
    | 'link'
    | 'link-external'
    | 'link-minor'
    | 'link-hover';

export type TextTypographyValue =
    | 'display-xl'
    | 'display-l'
    | 'display-m'
    | 'display-s'
    | 'headline-xl'
    | 'headline-l'
    | 'headline-s'
    | 'headline-xs'
    | 'headline-m'
    | 'subheader-xl'
    | 'subheader-l'
    | 'subheader-m'
    | 'subheader-s'
    | 'body-short-xl'
    | 'body-short-l'
    | 'body-short-m'
    | 'body-short-s'
    | 'body-long-xl'
    | 'body-long-l'
    | 'body-long-m'
    | 'body-long-s'
    | 'caption-xl'
    | 'caption-l'
    | 'caption-m'
    | 'overline-l'
    | 'overline-m'
    | 'overline-s'
    | 'control-xxs'
    | 'control-xs'
    | 'control-s'
    | 'control-l'
    | 'control-xl'
    | 'control-m'
    | 'control-xxl';

export interface TextProps {
    /**
     * Тип элемента для отображения как (строка или компонент).
     */
    as?: ElementType;

    /**
     * Основной контент
     */
    children?: ReactNode;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Задает выравнивание текста в компоненте.
     */
    align?: TextAlignValue;

    /**
     * Задает отображение переполненного текста
     */
    overflow?: TextOverflowValue;

    /**
     * Максимальное количество строк текста (работает в связке с overflow)
     */
    maxLines?: number;

    /**
     * Дополнительные стили
     */
    style?: CSSProperties;

    /**
     * Цвет текста
     */
    color?: TextColorValue;

    /**
     * Задает типографику текста в компоненте.
     */
    typography?: TextTypographyValue;
}

export const cnText = cn('Text');

/**
 * Базовый примитив представления текстовых данных.
 * @param { TextProps }  props
 */
export class Text extends PureComponent<TextProps> {
    static displayName = cnText();

    private static getLineHeight(textElement: HTMLElement) {
        return parseFloat(getComputedStyle(textElement).lineHeight);
    }

    private static isFade(overflow: TextOverflowValue) {
        return overflow === 'fade' || overflow === 'fade-horizontal';
    }

    private static isEllipsis(overflow: TextOverflowValue) {
        return overflow === 'ellipsis';
    }

    private textElement: HTMLElement | null = null;

    componentDidMount() {
        this.updateText();
    }

    componentWillUnmount() {
        this.textElement = null;
    }

    componentDidUpdate(prevProps: Readonly<TextProps>) {
        if (prevProps.typography !== this.props.typography) {
            this.updateText();
        }
    }

    private updateText() {
        const { textElement } = this;
        const { overflow } = this.props;
        if (textElement && overflow && Text.isFade(overflow) && canUseDOM()) {
            this.forceUpdate();
        }
    }

    private getLineHeight(): number | void {
        if (this.textElement) {
            const lineHeightValue = Text.getLineHeight(this.textElement);
            if (!isNaN(lineHeightValue)) {
                return lineHeightValue;
            }
        }
    }

    private setTextElement(el: HTMLElement) {
        this.textElement = el;
    }

    render() {
        const {
            as: ElementType = 'span',
            className,
            align,
            overflow,
            maxLines = 1,
            style = {},
            color,
            typography,
            ...otherProps
        } = this.props;

        if (overflow) {
            if (Text.isEllipsis(overflow)) {
                style.WebkitLineClamp = maxLines;
            } else if (Text.isFade(overflow)) {
                style.whiteSpace = maxLines === 1 ? 'nowrap' : undefined;
                if (maxLines > 1) {
                    const lineHeight = this.getLineHeight();
                    if (typeof lineHeight === 'number') {
                        style.maxHeight = `${maxLines * lineHeight}px`;
                    }
                }
            }
        }

        return (
            <ElementType
                ref={this.setTextElement.bind(this)}
                style={style}
                className={cnText(
                    {
                        color,
                        align,
                        overflow,
                    },
                    [className],
                )}
                {...otherProps}
            />
        );
    }
}
