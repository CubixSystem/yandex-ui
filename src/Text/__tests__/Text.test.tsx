import React from 'react';

import { createClientRender, screen } from '../../internal/testing';
import { Text, cnText } from '../Text';

describe('Text', () => {
    const render = createClientRender();

    test('должен быть установлен корректный displayName', () => {
        expect(Text.displayName).toBe(cnText());
    });

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(<Text className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен изменять тэг элемента через as', () => {
        render(<Text as="p" data-testid="element" />);

        expect(screen.getByTestId('element').tagName).toBe('P');
    });

    test.each`
        align
        ${'start'}
        ${'end'}
        ${'center'}
        ${'justify'}
    `('должен добавлять выравнивание `$align` через props', ({ align }) => {
        render(<Text align={align} data-testid="element" />);

        expect(screen.getByTestId('element')).toHaveClass(cnText({ align }));
    });

    test.each`
        overflow
        ${'ellipsis'}
        ${'fade'}
        ${'fade-horizontal'}
    `('должен управлять переполнением `$overflow` через props', ({ overflow }) => {
        render(<Text overflow={overflow} data-testid="element" />);

        expect(screen.getByTestId('element')).toHaveClass(cnText({ overflow }));
    });

    test('должен управлять цветом через props', () => {
        const color = 'primary';
        render(<Text color={color} data-testid="element" />);

        expect(screen.getByTestId('element')).toHaveClass(cnText({ color }));
    });
});
