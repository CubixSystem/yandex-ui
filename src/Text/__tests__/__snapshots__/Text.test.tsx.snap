// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text mix"
/>
`;
